### ISSUE TITLE
[Bug]: Redirecting synchronously/without flicker

### ISSUE THREAD

**Issue by mogelbrod** [remix-run/react-router#8295]
**Edit:** This might actually be a feature request - I wasn't sure if it's supposed to be possible or not.

### What version of React Router are you using?

react-router-dom@6.0.1

### Steps to Reproduce

1. Open https://codesandbox.io/s/react-router-v6-redirect-flash-5plmi?file=/src/App.js
2. Click on _Two_ (or _Three_) in the header
3. Click on _Root_ in the header

### Expected Behavior

The pages for _Root_, _Two_ and _Three_ are rendered immediately after clicking on their respective links.

### Actual Behavior

Clicking on _Root_ briefly renders nothing in the `<Routes>` block, followed by the correct content being shown shortly thereafter. This makes for a subpar experience.

I'm guessing this is due to `<Navigate>` triggering the redirect in an `useEffect()`. But I have not find any alternative ways to handle a redirect like this. react-router v3 had the [`onEnter` callback prop](https://github.com/remix-run/react-router/blob/v3/docs/API.md#onenternextstate-replace-callback) for `<Route>` which solved this by deferring the re-render until the callback completed, including allowing the next route to be modified.

**Relevant code:**
```jsx
<Routes>
  <Route index element={<Navigate replace to="one" />} />
</Routes>
...
<Link to="/" children="Root" />
```

**Comment by mjackson** [comment]
I wrote about how we are handling redirects on the initial render [in this gist](https://gist.github.com/mjackson/b5748add2795ce7448a366ae8f8ae3bb). Are you able to move your redirect to the server instead of doing it on the client?

A redirect on the client on the initial render is like a setState during render, which should be avoided if possible.

**Comment by mogelbrod** [comment]
Appreciate the quick reply @mjackson!
This is for a (currently) basic client-side rendered app with no server side redirects.
Note that the flicker happens every time the user navigates to `/` using the navigation links, not only the initial render.

**Comment by iMoses-Apiiro** [comment]
@mjackson I've been working on pure single page applications for many years now, and I think this use case is very common in such cases. If we're considering that routes can be registered in multiple places it's reasonable that a top route will redirect to somewhere like `/profile` while the profile page has multiple tabs and we want to redirect to the first one in case we're at the index. 

The flickering is a regression from what we have today. It'd be nice if this special case of redirecting directly from the routes definitions (without using `element`) would be officially addressed.

**Comment by nathan-knight** [comment]
In my case in my application's navigation menu I use a `NavLink` to a parent route so it will match child routes and I automatically redirect from the parent route to a child route. This trick is necessary because a `NavLink` can't match one route and link to another.

**Comment by kresli** [comment]
Same here, not able to move redirect to server. It also doesn't make much sense for me as I would need to keep server and client routes in sync. I will probably switch back to the previous version though.

**Comment by mogelbrod** [comment]
@mjackson is there any possibility of react-router providing a way to redirect without the useEffect? Are there any additional things the community could help out with?

**Comment by slk333** [comment]
This is still a big issue.
The suggestion to move the logic to the server is largely irrelevant for people who maintain create-react-app web apps. React-router never said it was not compatible with client side routing as far as I know, so I feel disappointed that the maintainer have such low regard to people who do client side routing

**Comment by hichemfantar** [comment]
I thought there was something wrong with my routing logic but apparently this is an issue with v6.
I agree that this is a regression compared to v5.

**Comment by brophdawg11** [comment]
Hey folks!  Redirecting _during render_ like this is essentially an issue because routing and rendering are tightly coupled.  One of the primary drivers of the new Data APIs released in 6.4.0 is to decouple routing/data loading from rendering, so the way to get rid of this flicker is to redirect in a loader: https://codesandbox.io/s/react-router-v6-redirect-flash-forked-qzkw87.

**Comment by hichemfantar** [comment]
Looks like the loader pattern fixes the issue. Thanks @brophdawg11 🎉
